# **STATS 503, Group Work Assignment 5: Cross validation**

**Instructions:** During lab section, and afterward as necessary, you will collaborate in two-person teams (assigned by the GSI) to complete the problems that are interspersed below. The GSI will help individual teams encountering difficulty, make announcements addressing common issues, and help ensure progress for all teams. **During lab, feel free to flag down your GSI to ask questions at any point!** Upon completion, one member of the team should submit their team's work through Canvas as html.

**credits:** Roman Kouznetsov, adapted from notes from Gang Qiao.

# Getting started



In this assignment, we will explore logistic regression and $k$-fold cross validation. In $k$-fold cross-validation, we partition a dataset into $k$ equally sized non-overlapping subsets $S$. For each subset $S_i$, a model is trained on $S\backslash S_i$ and evaluated on $S_i$. The cross-validation estimator of the prediction's error is the average of the prediction errors obtained on each fold.

Let's start with simple example that will help us understand how CV works. We'll first import the relevant packages.

In [108]:
from sklearn.datasets import make_classification
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.model_selection import train_test_split
import sklearn.metrics
import numpy as np
import pandas as pd

For demonstration purposes, let's create a synthetic binary classification dataset using scikit-learn's `make_classification` function.

In [109]:
# Generate a synthetic binary classification dataset
X, y = make_classification(n_samples=1000, n_features=20, n_classes=2, random_state=42)

We'll look at two ways to do cross-validation.  First we'll do it "by hand." Then we'll show how sklearn makes it easy.

# K-fold CV by hand

First, we divide the data into 10 parts (a.k.a. folds).

In [110]:
n_splits = 10 # 10-fold
rng = np.random.default_rng(0) # make a random number generator with fixed seed
permutation = rng.permutation(len(X)) # create a shuffling of the indices of our data
splits = np.split(permutation,n_splits) # make folds

In [111]:
# e.g., these are the *indices* of the datapoints "fold 3"
splits[3]

array([290, 277, 360, 524, 462, 789, 908, 493, 818, 251, 623, 123, 781,
       589, 874, 304, 869, 746, 576, 811, 433, 837,  37, 161,  58, 560,
       643, 965, 163,  64, 475, 838, 168, 659, 109,  41, 756, 596, 303,
       763, 101, 703, 120, 531, 366, 553, 548, 575, 731, 470, 578, 660,
       551, 417, 716, 911, 227, 587, 872, 185, 513, 866, 321, 418, 760,
       820, 119, 952,  24, 606, 170, 278, 514,  15, 155, 367, 828,  92,
        55, 897, 118, 791, 701, 145, 860, 645, 620, 778, 497, 794, 246,
       842,  80, 646, 225, 951, 482,  20, 343, 250])

In [112]:
# and these are the y values for the corresponding samples
y[splits[3]]

array([1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1,
       1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
       1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0,
       0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0])

Now, we'll assess predictive performance 10 times.  Each time we'll hold out a different fold and use the rest of the data for training.

In [113]:
# we'll look at two metrics of predictive performance,
# and store the results in these arrays
missclass = np.zeros(len(splits))
aurocs = np.zeros(len(splits)) # since this is binary classification, we can assess via auroc

# iterate through folds
for i in range(len(splits)):
    # create test/train split for this held-out fold
    s = list(splits) # copy list
    test = s.pop(i) # pop out the ith fold
    training = np.concatenate(s) # combine all remaining data for test

    # fit model
    model = LogisticRegression() # make estimator
    model.fit(X[training], y[training]) # fit estimator using the training data

    # assess metric for predictive performance using test data
    missclass[i] = np.mean(model.predict(X[test])!= y[test]) # get misclassification on test data
    aurocs[i] = sklearn.metrics.roc_auc_score(y[test], model.predict_proba(X[test])[:, 1]) # get auroc on test data

In [114]:
missclass # misclassification rate for each of the 10 test/train splits

array([0.15, 0.15, 0.16, 0.1 , 0.21, 0.13, 0.11, 0.09, 0.15, 0.12])

In [115]:
aurocs # auroc for each of the 10 test/train splits

array([0.89783654, 0.90381494, 0.91304348, 0.93427036, 0.91666667,
       0.90868506, 0.9696    , 0.96025692, 0.93830128, 0.95392628])

In [116]:
# report conclusions: mean(scores) pm sqrt(tau^2/K)
print(f"Mean misclassification rate: {missclass.mean():.4f}, plus minus {np.sqrt(missclass.var()/n_splits):.4f}")
print(f"Auroc: {aurocs.mean():.4f}, plus minus {np.sqrt(aurocs.var()/n_splits):.4f}")

Mean misclassification rate: 0.1370, plus minus 0.0105
Auroc: 0.9296, plus minus 0.0076


# K-fold CV using `sklearn.model_selection`

In [117]:
model = LogisticRegression()

Setup the $k$-fold cross-validation configuration. For example, using 10 folds...
* `n_splits=10` means 10-fold
* `shuffle=True` meaning random folds
* `random_state=42` means we'll get the same random folds each time

In [118]:
cv = KFold(n_splits=10, random_state=42, shuffle=True)

In [119]:
# look at folds generated by this cross-validation splitter
for i, (train, test) in enumerate(cv.split(X)):
    # this automatically gives you the train indices and test indices
    # without having to construct them yourselves by combining folds
    print(f'\nsplit {i}')
    print('   test:   [', ' '.join(test[:10].astype('U')),'...]') # test indices
    print('   train:  [', ' '.join(train[:10].astype('U')),'...]') # train indices


split 0
   test:   [ 10 23 30 39 54 59 63 66 67 70 ...]
   train:  [ 0 1 2 3 4 5 6 7 8 9 ...]

split 1
   test:   [ 25 44 55 60 72 78 86 110 120 137 ...]
   train:  [ 0 1 2 3 4 5 6 7 8 9 ...]

split 2
   test:   [ 2 3 5 7 9 29 31 33 49 65 ...]
   train:  [ 0 1 4 6 8 10 11 12 13 14 ...]

split 3
   test:   [ 0 6 28 41 56 69 73 79 90 108 ...]
   train:  [ 1 2 3 4 5 7 8 9 10 11 ...]

split 4
   test:   [ 11 12 18 24 42 43 51 61 74 83 ...]
   train:  [ 0 1 2 3 4 5 6 7 8 9 ...]

split 5
   test:   [ 15 19 22 38 46 50 57 68 75 93 ...]
   train:  [ 0 1 2 3 4 5 6 7 8 9 ...]

split 6
   test:   [ 8 16 17 26 36 37 45 48 53 103 ...]
   train:  [ 0 1 2 3 4 5 6 7 9 10 ...]

split 7
   test:   [ 112 122 123 129 143 146 147 183 186 197 ...]
   train:  [ 0 1 2 3 4 5 6 7 8 9 ...]

split 8
   test:   [ 4 14 27 32 35 40 47 52 62 64 ...]
   train:  [ 0 1 2 3 5 6 7 8 9 10 ...]

split 9
   test:   [ 1 13 20 21 34 58 71 80 87 91 ...]
   train:  [ 0 2 3 4 5 6 7 8 9 10 ...]


In fact, you don't even have to run the for loop yourself!

With a single call to cross_val_score, we can evaluate the model using cross-validation. Here, we'll use accuracy as the performance metric, but you can choose other metrics like precision, recall, etc.

In [120]:
# Calculate accuracies across folds
scores = cross_val_score(model, X, y, scoring='accuracy', cv=cv, n_jobs=-1)
scores

array([0.87, 0.86, 0.83, 0.89, 0.88, 0.86, 0.85, 0.87, 0.9 , 0.86])

In [121]:
print(f"Mean accuracy: {scores.mean():.4f}, plus minus {np.sqrt(scores.var()/len(scores)):.4f}")
print(f"Mean misclassification rate: {1-scores.mean():.4f}, plus minus {np.sqrt(scores.var()/len(scores)):.4f}")

Mean accuracy: 0.8670, plus minus 0.0060
Mean misclassification rate: 0.1330, plus minus 0.0060


Note that conclusions are not exactly the same, because CV used different splits.

If you want to use multiple metrics, use skelarn.cross_validate instead.

In [122]:
# info for each split
results = pd.DataFrame(
    sklearn.model_selection.cross_validate(model, X, y, cv=cv,
                                           scoring=('accuracy','roc_auc'),
                                           return_train_score=True)
    )
results

fit_time  score_time  test_accuracy  train_accuracy  test_roc_auc  \
0  0.017527    0.005358           0.87        0.874444      0.910879   
1  0.017630    0.005205           0.86        0.880000      0.926250   
2  0.004290    0.004195           0.83        0.881111      0.900641   
3  0.004339    0.004414           0.89        0.872222      0.942222   
4  0.004392    0.004926           0.88        0.877778      0.951600   
5  0.004527    0.009968           0.86        0.876667      0.939382   
6  0.004395    0.004373           0.85        0.883333      0.910364   
7  0.005307    0.004638           0.87        0.880000      0.890405   
8  0.005234    0.004553           0.90        0.870000      0.961851   
9  0.005265    0.005287           0.86        0.876667      0.958788   

   train_roc_auc  
0       0.937207  
1       0.936828  
2       0.939317  
3       0.934140  
4       0.933526  
5       0.934303  
6       0.937373  
7       0.939992  
8       0.932240  
9       0.932327

In [123]:
print(f"Auroc: {results.test_roc_auc.mean():.4f}, plus minus {np.sqrt(results.test_roc_auc.var()/len(results)):.4f}")
print(f"Accuracy: {results.test_accuracy.mean():.4f}, plus minus {np.sqrt(results.test_accuracy.var()/len(results)):.4f}")

Auroc: 0.9292, plus minus 0.0080
Accuracy: 0.8670, plus minus 0.0063


Note the results also include train performance (because we set `return_train_score=True`).  Can sometimes be interesting to see discrepancy between train and test performance.  Usually train performance is a bit better.  If train performance is *a lot* better your estimator may have "too much flexibility" (though sometimes you may also be experiencing so-called "benign overfitting" in which case your estimator is actually just fine...).  

In [124]:
print(f"Train auroc: {results.train_roc_auc.mean():.4f}, plus minus {np.sqrt(results.train_roc_auc.var()/len(results)):.4f}")
print(f"Train accuracy: {results.train_accuracy.mean():.4f}, plus minus {np.sqrt(results.train_accuracy.var()/len(results)):.4f}")

Train auroc: 0.9357, plus minus 0.0009
Train accuracy: 0.8772, plus minus 0.0013


# Compare with another estimator (using same splits!)

In [125]:
model2 =  KNeighborsClassifier(5)
scores2 = cross_val_score(model2, X, y, scoring='accuracy', cv=cv,n_jobs=-1) # note, using same folds, cv

print(f"Mean Accuracy: {scores2.mean():.4f}, plus minus {np.sqrt(scores2.var()/len(scores2)):.4f}")

Mean Accuracy: 0.8120, plus minus 0.0110


It seems that 5-NN is worse, by a margin that is well in excess of the spread.

# Bias-variance tradeoff

It is very helpful to think about bias-variance tradeoff in cross-validation. In CV, the number of folds to use (the value of $k$) is an important decision. Imagine repeating the learning procedure on multiple datasets. The lower the value for $k$, the higher the bias in the error estimates and the less variance **across datasets**. Conversely, when $k$ is set equal to the training+val sample size, the error estimate is then very low in bias but has the possibility of high variance **across datasets**.

Why? Some intuitions (but not mathematically rigorous proof) here:

While there is no overlap between the test sets on which the models are evaluated, there is overlap between the training sets for all $k>2$. The overlap is largest for leave-one-out cross-validation. This means that the learned models are correlated, i.e. dependent, and the variance of the sum of correlated variables increases with the amount of covariance:
$
Var(\sum_i X_i) = \sum_i \sum_j Cov(X_i, X_j).
$
Therefore, leave-one-out cross-validation has large variance in comparison to CV with smaller $k$. To summarize, larger $k$ means less bias towards overestimating the true expected error (as training folds will be closer to the total dataset) but higher variance and higher running time (as you are getting closer to the limit case: Leave-One-Out CV).

For more fun facts and simulation about bias-variance tradeoff and cross validation, please see [this post](https://stats.stackexchange.com/questions/61783/bias-and-variance-in-leave-one-out-vs-k-fold-cross-validation/357749#357749).

# Group Work 5 (30 points)

But WHY do we even bother with cross-validation? What is the point?

In this group work assignment, you'll perform K-fold cross validation on several classification models on the NHANES dataset. Additionally, you will be asked to write answers to explain WHY we do the things we do. Please feel free to ask instructors for guidance if cross-validation is still new to you.



---


### Part 1: Dataset Setup (1 point)

Our favorite (and only) dataset we have used in the group work assignments is back! Please take the appropriate time to load in the 3 NHANES datasets.

In [126]:
hdl = pd.read_sas('HDL_L.xpt')
bmx = pd.read_sas('BMX_L.xpt')
demo = pd.read_sas('DEMO_L.xpt')
df = pd.merge(hdl, bmx, on = 'SEQN')
df = pd.merge(df, demo, on = 'SEQN')



---


### Part 2: Variable Setup and Selection (3 points)

For this problem, we will again try to predict individuals that have high-density lipoprotein (HDL) cholesterol of greater than 60. An HDL of 60 **mg/dL** or higher is often viewed as protective against heart disease—this is typically the level you’d like to aim for, if possible. For this task please do the following:

1. Create the binary indicator variable. Also, use the following features for predictive purposes: Gender, Age, Weight, Height, BMI, WaistSize, Household Size, and Ethnicity. You may need to refer to the docs to figure out their variable names.



*   [HDL_L](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2021/DataFiles/HDL_L.htm)
*   [DEMO_L](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2021/DataFiles/DEMO_L.htm)
*   [BMX_L](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2021/DataFiles/BMX_L.htm)



2. **Please** change the variable names to be English-legible but still python variable style.

3. Drop all missing values.

In [127]:
df.rename(columns={'RIAGENDR': 'Gender', 'RIDAGEYR': 'Age', 'BMXWT': 'Weight',
                   'BMXHT': 'Height', 'BMXBMI': 'BMI', 'BMXWAIST': 'WaistSize',
                   'SDMVPSU': 'Household_Size', 'RIDRETH1': 'Ethnicity'}, inplace=True)

my_df = df.copy()
my_df['HDL_60'] = (my_df['LBDHDD'] > 60).astype(int)

my_df = my_df[['Gender', 'Age', 'Weight', 'Height', 'BMI', 'WaistSize', 'Household_Size', 'Ethnicity', 'HDL_60']]

my_df = my_df.dropna()

In [128]:
assert my_df.shape[0] / df.shape[0] > 0.95



---


### Part 3: Training + Testing Split (1 point)

Please split your data into a train and test set, with 70% of observations in the train set and 30% in the test set.

*   Please use the `train_test_split` method.
*   Please use the `random_state=42`.
*   Please stratify the sampling to include roughly the same distribution of response values in each set. [Why should we stratify?](https://scikit-learn.org/stable/modules/cross_validation.html#stratified-k-fold)


In [129]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)



---


### Part 4: Implementing K-fold CV (2 points)

Write a function called KFoldCV that takes in as input 4 arguments:

1.   `X` The predictors array.
2.   `y` The response variable array.
3.   `model`: An sklearn model object on which we can call fit and predict.
4.   `K`: An integer representing the number of folds we want to include.

and returns the list of classification accuracies for each fold.



In [130]:
from sklearn.metrics import accuracy_score
def KFoldCV(X, y, model, K=10):
  cv = KFold(n_splits=K, shuffle=True, random_state=42)
  scores = cross_val_score(model, X, y, scoring='accuracy', cv=cv, n_jobs=-1)
  return scores



---


### Part 5 (Code): Execution (2 points)

Run your KFoldCV function using Logistic Regression with $K=10$.

Please ensure the that the output of the function is visible in the notebook.

<u>Note</u>: In the event you get a warning that the model has not reached convergence, add the argument `max_iter=1000` to the LogisticRegression instance.

In [131]:
model = LogisticRegression()
KFoldCV(X_train, y_train, model, K=10)

array([0.92857143, 0.8       , 0.88571429, 0.82857143, 0.84285714,
       0.88571429, 0.87142857, 0.84285714, 0.87142857, 0.92857143])

### Part 5 (Writing): Generalization (2 points)

One motivation for cross-validation is to assess the generalizability of your model on unseen data.

Explain in at most two sentences whether you believe your model generalizes well on unseen data. Reference the output of your function to make your case.

**ANSWER:**

$$\boxed{\textrm{
The cross-validation output shows that our model consistently achieves accuracy. it is likely to generalize well on unseen data.
}}$$



---


### Part 6 (Code): Working with Regularization (3 points)

So turns out, if you look at the LogisticRegression module in sklearn, it uses an L2 penalty by default! You can see for yourself on the [documentation page for sklearn.linear_model.LogisticRegression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html).

One problem, we just accepted the default regularization term: C=1.0. How do we know this was the right way to go? Let's see if it even is.

Write a function that performs K-Fold validation but with each fold having a different value for $C$. The values considered should range be $10^{-5}, 10^{-4}, \dots 10^{4}$.

In [132]:
from sklearn.metrics import accuracy_score
# function returns dictionary
def KFoldCV_L2(X, y, K=10) -> dict:
    C_values = [10**i for i in range(-5, 5)]
    results = {}

    kf = KFold(n_splits=K, shuffle=True, random_state=42)

    for C in C_values:
        model = LogisticRegression(C=C, max_iter=1000)
        scores = cross_val_score(model, X, y, scoring='accuracy', cv=kf, n_jobs=-1)
        results[C] = scores

    return results

In [133]:
KFoldCV_L2(X_train, y_train)

{1e-05: array([0.41428571, 0.51428571, 0.5       , 0.47142857, 0.72857143,
        0.51428571, 0.7       , 0.71428571, 0.47142857, 0.51428571]),
 0.0001: array([0.58571429, 0.72857143, 0.8       , 0.7       , 0.84285714,
        0.74285714, 0.82857143, 0.9       , 0.84285714, 0.85714286]),
 0.001: array([0.9       , 0.81428571, 0.88571429, 0.78571429, 0.84285714,
        0.91428571, 0.85714286, 0.87142857, 0.9       , 0.91428571]),
 0.01: array([0.91428571, 0.82857143, 0.91428571, 0.8       , 0.84285714,
        0.91428571, 0.88571429, 0.87142857, 0.88571429, 0.92857143]),
 0.1: array([0.92857143, 0.8       , 0.9       , 0.81428571, 0.84285714,
        0.88571429, 0.88571429, 0.84285714, 0.87142857, 0.92857143]),
 1: array([0.92857143, 0.8       , 0.88571429, 0.82857143, 0.84285714,
        0.88571429, 0.87142857, 0.84285714, 0.87142857, 0.92857143]),
 10: array([0.92857143, 0.81428571, 0.88571429, 0.82857143, 0.84285714,
        0.88571429, 0.87142857, 0.84285714, 0.87142857, 0.928571

### Part 6 (Writing): Evaluation (2 points)

Based on your cross-validation, do you believe that regularization plays an effect on the predictive accuracy on unseen accuracy? What level of regularization should we choose?

**ANSWER:**

$$\boxed{\textrm{
Yes. As *C* becomes larger (i.e., penalizing larger numbers less), the accuracy scores change.  A moderate level of regularization appears optimal
}}$$



---


### Part 7 (Code): Train+Val+Test (3 points)

Please now retrain your final model (with the best regularization value you found in the last part) on all training data. Then, evaluate your model's performance on the test set.

In [134]:
model = LogisticRegression(C = 0.01)
KFoldCV(X_train, y_train, model, K=10)

array([0.91428571, 0.82857143, 0.91428571, 0.8       , 0.84285714,
       0.91428571, 0.88571429, 0.87142857, 0.88571429, 0.92857143])

### Part 7 (Writing 1): Retraining on Training + Validation Data (1 point)

Explain in 1-2 sentences **MAX** why we retrain the model on the merged train+val dataset.

**ANSWER:**

$$\boxed{\textrm{
We want the model to learn from all available data, leading to a more accurate and reliable estimation of parameters.
}}$$

### Part 7 (Writing 2): Evalution (1 point)

Is logistic regression performing well on this data? Explain in 1-2 sentences **MAX**. **You may need some supporting code for your argument**.

<u>Hint</u>: Think about what a trivial baseline would be.

In [135]:
from sklearn.dummy import DummyClassifier
dummy = DummyClassifier(strategy="most_frequent")
dummy.fit(X_train, y_train)
baseline_accuracy = dummy.score(X_test, y_test)
print("Trivial baseline accuracy:", baseline_accuracy)

Trivial baseline accuracy: 0.48333333333333334


**ANSWER:**

$$\boxed{\textrm{
The logistic regression is doing much better than a plain dummy classifier.
}}$$



---


### Part 8: K-Folds on K-Nearest Neighbors (2 points)

Using your KFoldCV function you created in Part 4, please run CV on 2-NN classification.

In [136]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=2)

knn_scores = KFoldCV(X, y, knn, K=10)

print("2-NN CV accuracies:", knn_scores)

2-NN CV accuracies: [0.7  0.78 0.75 0.78 0.74 0.8  0.78 0.84 0.78 0.79]




---


### Part 9: Finding the Right K (4 points)

Similar to what you did for regularized logistic regression, please run a 10-fold CV that assesses model performance on 10 different number of neighbors: [1, 3, 5, 10, 15, 20, 25, 35, 50, 100]. You should attempt to predict each fold with each num_neighbors value.

Please write a function called `KFoldCV_NN` that takes as input `X`, `y`, and `K` and returns a dictionary. The dictionary should have the number of neighbors considered as keys and the mean validation accuracy as values.



In [137]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score

def KFoldCV_NN(X, y, K=10):
    num_neighbors = [1, 3, 5, 10, 15, 20, 25, 35, 50, 100]
    results = {}  # Initialize dictionary outside the loop
    for k in num_neighbors:
        model = KNeighborsClassifier(n_neighbors=k)
        scores = cross_val_score(model, X, y, scoring='accuracy', cv=K, n_jobs=-1)
        results[k] = scores.mean()
    return results


In [138]:
KFoldCV_NN(X_train, y_train)

{1: 0.7742857142857142,
 3: 0.8242857142857142,
 5: 0.82,
 10: 0.8357142857142857,
 15: 0.8585714285714288,
 20: 0.8614285714285715,
 25: 0.8528571428571429,
 35: 0.8571428571428571,
 50: 0.8514285714285714,
 100: 0.8614285714285714}



---


### Part 10: Retraining on Training + Validation Data (2 points)

Please retrain your data on both the training and validation set. Make sure you run the KNeighborsClassifier with the best number of neighbors as you determined from the previous part.

In [139]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

best_knn = KNeighborsClassifier(n_neighbors=20)

best_knn.fit(X_train, y_train)

predictions = best_knn.predict(X_test)
test_accuracy = accuracy_score(y_test, predictions)
print("Test Accuracy:", test_accuracy)


Test Accuracy: 0.8166666666666667




---


### Part 11: Test Set Evaluation (1 point)


Evaluate your final model's performance.

The final KNN model with 20 neighbors achieves a test accuracy of approximately 81.67%, which suggests that it generalizes reasonably well on unseen data and performs better than a trivial baseline